### PR TITLE
Add unit tests for pipeline() - issue 95

### PR DESCRIPTION
**Pull request by AndresSanchezCuadrado**
Unit tests pipeline

### CODE REVIEW

File: unit_tests/pipeline.cpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+#include "supported_executions.h"
+
+using namespace std;
+using namespace grppi;
+
+template <typename T>
+class pipeline_test : public ::testing::Test {
+public:
+  T execution_;
+  polymorphic_execution poly_execution_ =
```
Comment (jdgarciauc3m) [inline unit_tests/pipeline.cpp:40]: Remove poly execution testing. We will handle this as part of issue-90

File: unit_tests/pipeline.cpp [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    },
+    [this]( auto x ) {
+      this->invocations_last++;
+    }
+  );
+
+  this->check_empty();
+}
+
+TYPED_TEST(pipeline_test, poly_empty)
```
Comment (jdgarciauc3m) [inline unit_tests/pipeline.cpp]: Remove all the poly_* test cases. We will add them as part of issue #90

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Remove anything related to polymorphic_execution from this test. That will be handled separately with issue #90

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Reorganize test cases.

As a minimum I see:

* Pipeline with 0 stages (zero_stages)
* Pipeline with 1 stage (one_stage)
* Pipeline wiht 2 stages (two_stages)
* Pipeline with 3 stages (three_stages)

**Comment by AndresSanchezCuadrado** [comment]
For this unit test to work, the issue #176 needs to be merged with latest.

File: unit_tests/pipeline.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        if(this->counter  <= 0){
+          return optional< std::vector<int> >();
+        }else{
+          return optional<std::vector<int>>(v);
+        }
+    },
+    grppi::farm(this->execution_,
+        [this](std::vector<int> v) {
+          //this->invocations_intermediate++;
```
Comment (AndresSanchezCuadrado) [inline unit_tests/pipeline.cpp]: Problem with scope. If this is uncommented a segmentation fault is produced.

Comment (jdgarciauc3m) [reply unit_tests/pipeline.cpp]: This needs to be debugged and solved.

A core is generated with native and TBB.

**Review by jdgarciauc3m — APPROVED** [review]
OK

**Comment by jdgarciauc3m** [comment]
Solves issue #95

**Pull request merged**